File: src/Santa.hs
module Santa (calculateKidID, findSolution) where

import Data.Tuple


-- Find the first kid's ID in any given sack
findFirst :: Integer -> Integer
findFirst 1    = 1
findFirst sack = sack ^ 2 `quot` 2

-- Calculate the kid's ID for given sack and present pair
calculateKidID :: (Integer, Integer) -> Integer
calculateKidID (sack, present) =
  let
      initialSquare = if odd sack && sack > 1 then sack - 1 else sack
      finalSquare = initialSquare + present - 1
      sign = if odd $ initialSquare + finalSquare then 1 else -1
      summedSquares = sumTo finalSquare + sign * sumTo initialSquare
  -- From the sum of squares we need to either subtract or add the first kid
  -- depending on whether the present is odd or even
  in summedSquares - (1 - 2 * (present `rem` 2)) * findFirst sack
  where
    -- Sum numbers from 1 to n, e.g. for n = 3, then return 1 + 2 + 3 = 6
    sumTo n = (1 + n) * n `quot` 2

calculateFactors :: Integer -> [(Integer, Integer)]
calculateFactors n =
  -- Add the opposite pair of factors, e.g. (1, 10) => (10, 1)
  factorPairs ++ map swap (filter (uncurry (/=)) factorPairs)
  where
    -- x divisors of n up to where x^2 >= n, rest is calculated by swapping pairs
    divisors = filter ((==0) . rem n) (takeWhile ((<=n) . (^2)) [1..n])
    -- The list of factor pairs, e.g. if n == 10 then (1, 10), (2, 5)
    factorPairs = (\sack -> (sack, n `quot` sack)) <$> divisors

findSolution :: Integer -> Integer
findSolution n =
  sum (calculateKidID <$> calculateFactors n) `rem` 10^8

File: src/santa.py
def find_first(sack):
    """ Find the ID of the first kid who gets allocated a present from the supplied sack. """
    return 1 if sack == 1 else int(sack ** 2 / 2)


def calculate_kid_id(sack, present):
    """ Find the ID of the kid receiving the supplied present in the sack """
    from_square = sack - 1 if sack % 2 == 1 and sack != 1 else sack
    to_square = from_square + present - 1
    sum_of_squares_after = (1 + to_square) * to_square // 2
    sum_of_squares_prior = (from_square + 1) * from_square // 2

    sign = 1 if (from_square + to_square) % 2 else -1

    s = sum_of_squares_after + sign * sum_of_squares_prior

    return s + (1 if present % 2 else -1) * find_first(sack)


def calculate_factors(x):
    """ Find all pairs of integers whose product equals x. """
    factors = [(sack, int(x / sack)) for sack in range(1, int(x ** 0.5) + 1) if (float(x) / sack).is_integer()]
    return factors + [(x[1], x[0]) for x in factors if x[0] != x[1]]


def calculate_allocation(n):
    """ Find the hash of the kid IDs for the provided input. """
    return [(factor, calculate_kid_id(*factor)) for factor in calculate_factors(n)]


def find_solution(n):
    """ Calculate the checksum as requested by Santa """
    return sum(a[1] for a in calculate_allocation(n)) % 100000000
